src/main/java/com/google/api/codegen/DiscoveryContext.java

```diff
@@ -84,6 +84,108 @@ public String getSimpleName(String name) {
     return name.substring(name.lastIndexOf('.') + 1);
   }
 
+  public String getSampleVarName(String typeName) {
+    return upperCamelToLowerCamel(getSimpleName(typeName));
+  }
+
+  /**
+   * Returns a name for a type's field's type.
+   */
+  public String typeName(Type type, Field field, String name) {
+    if (field.getCardinality() == Field.Cardinality.CARDINALITY_REPEATED) {
+      if (isMapField(type, field.getName())) {
+        return mapTypeName(field);
+      } else {
+        return arrayTypeName(field);
+      }
+    } else {
+      if (field.getKind() == Field.Kind.TYPE_MESSAGE) {
+        return objectTypeName(field);
+      } else {
+        return nativeTypeName(type, field, name);
+      }
+    }
+  }
+
+  /**
+   * Returns a name for an array field's type.
+   */
+  protected String arrayTypeName(Field field) {
+    return arrayTypeName(elementTypeName(field));
+  }
+
+  protected String arrayTypeName(String elementName) {
+    return String.format("%s_array", elementName);
```

Comment:
Is this always idiomatic, as opposed to declaring the array type inline? Same below. Please comment if meant to be overriden by each language.